### PR TITLE
Move all query methods to RealmQuery

### PR DESCRIPTION
**Pull request by cmelchior**
Part of #1594. All changes there should be merged in one release

@realm/java 

Final commit message:
-----
Title: Deprecate allObjects and distinct methods on Realm/DynamicRealm

Currently we have the following helper query methods on Realm/DynamicRealm: `allObjects()` and `distinct()`

However we have to be mindful about the number of methods in our API and from the projects we have seen so far, these methods does not seem to have been used much.

Moving all query methods to RealmQuery has two advantages: 1) It helps us to reduce our overall method count and 2) It makes our API more consistent as now all RealmQuery methods are only found on RealmQuery.

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -1,3 +1,12 @@
+## 0.90.0
+
+### Deprecated
+
+* `Realm.allObjects*()`. Use `Realm.where(clazz).findAll*()` instead.
+* `Realm.distinct*()`. Use `Realm.where(clazz).distinct*()` instead.
+* `DynamicRealm.allObjects*()`. Use `Realm.where(className).findAll*()` instead.
```
Comment (zaki50) [inline CHANGELOG.md]: `Realm.where` should be `DynamicRealm.where`

File: CHANGELOG.md [hunk 2]
```diff
@@ -1,3 +1,12 @@
...
+* `DynamicRealm.distinct*()`. Use `Realm.where(className).distinct*()` instead.
```
Comment (zaki50) [inline CHANGELOG.md]: same

File: realm/realm-library/src/main/java/io/realm/Realm.java [hunk 3]
```diff
@@ -1030,16 +1003,9 @@ boolean contains(Class<? extends RealmObject> clazz) {
...
-     * Gets all objects of a specific Class sorted by multiple fields. If no objects exist, the returned
-     * {@link RealmResults} will not be null. The RealmResults.size() to check the number of objects instead.
-     *
-     * @param clazz the Class to get objects of.
-     * @param sortOrders sort ascending if Sort.ASCENDING, sort descending if Sort.DESCENDING.
-     * @param fieldNames an array of field names to sort objects by. The objects are first sorted by fieldNames[0], then
-     *                   by fieldNames[1] and so forth.
-     * @return a sorted RealmResults containing the objects.
-     * @throws java.lang.IllegalArgumentException if a field name does not exist.
+     * DEPRECATED: Use {@code realm.where(clazz).findAllSorted(fieldNames[], sortOrders[])` instead.
```
Comment (zaki50) [inline realm/realm-library/src/main/java/io/realm/Realm.java]: `{@code ` is not closed.

File: realm/realm-library/src/main/java/io/realm/Realm.java [hunk 4]
```diff
@@ -1008,19 +990,10 @@ boolean contains(Class<? extends RealmObject> clazz) {
...
-     * @param clazz the class ti get objects of.
-     * @param fieldName1 first field name to sort by.
-     * @param sortOrder1 sort order for first field.
-     * @param fieldName2 second field name to sort by.
-     * @param sortOrder2 sort order for second field.
-     * @param fieldName3 third field name to sort by.
-     * @param sortOrder3 sort order for third field.
-     * @return a sorted RealmResults containing the objects.
-     * @throws java.lang.IllegalArgumentException if a field name does not exist.
+     * DEPRECATED: Use {@code realm.where(clazz).findAllSorted(fieldName1, sortOrder1, fieldName2, sortOrder2, fieldName3, sortOrder3)`
```
Comment (zaki50) [inline realm/realm-library/src/main/java/io/realm/Realm.java]: `{@code ` is not closed.

File: realm/realm-library/src/main/java/io/realm/Realm.java [hunk 5]
```diff
@@ -989,17 +979,9 @@ boolean contains(Class<? extends RealmObject> clazz) {
...
-     * {@link RealmResults} will not be {@code null}. The RealmResults.size() to check the number of objects instead.
-     *
-     * @param clazz the class ti get objects of.
-     * @param fieldName1 first field name to sort by.
-     * @param sortOrder1 sort order for first field.
-     * @param fieldName2 second field name to sort by.
-     * @param sortOrder2 sort order for second field.
-     * @return a sorted RealmResults containing the objects.
-     * @throws java.lang.IllegalArgumentException if a field name does not exist.
+     * DEPRECATED: Use {@code realm.where(clazz).findAllSorted(fieldName1, sortOrder1, fieldName2, sortOrder2)` instead.
```
Comment (zaki50) [inline realm/realm-library/src/main/java/io/realm/Realm.java]: `{@code ` is not closed.

File: realm/realm-library/src/main/java/io/realm/Realm.java [hunk 6]
```diff
@@ -953,27 +953,17 @@ boolean contains(Class<? extends RealmObject> clazz) {
...
     /**
-     * Get all objects of a specific Class sorted by a field. If no objects exist, the returned {@link RealmResults}
-     * will not be {@code null}. The RealmResults.size() to check the number of objects instead.
-     *
-     * @param clazz the Class to get objects of.
-     * @param fieldName the field name to sort by.
-     * @param sortOrder how to sort the results.
-     * @return a sorted RealmResults containing the objects.
-     * @throws java.lang.IllegalArgumentException if field name does not exist.
+     * DEPRECATED: Use {@code realm.where(clazz).findAllSorted(fieldName, sortOrder)` instead.
```
Comment (zaki50) [inline realm/realm-library/src/main/java/io/realm/Realm.java]: `{@code ` is not closed.

File: realm/realm-library/src/main/java/io/realm/Realm.java [hunk 7]
```diff
@@ -953,27 +953,17 @@ boolean contains(Class<? extends RealmObject> clazz) {
     }
 
     /**
-     * Gets all objects of a specific Class. If no objects exist, the returned RealmResults will not be {@code null}.
-     * The RealmResults.size() to check the number of objects instead.
-     *
-     * @param clazz the Class to get objects of.
-     * @return a RealmResult list containing the objects.
-     * @see io.realm.RealmResults
+     * DEPRECATED: Use {@code realm.where(clazz).findAll()` instead.
```
Comment (zaki50) [inline realm/realm-library/src/main/java/io/realm/Realm.java]: `{@code ` is not closed.

**Comment by kneth** [comment]
I know that `RealmQuery.count()` is fast but how often is it needed? It might be better to remove it and use `realm.where(clz).findAll().size()` instead?

**Comment by cmelchior** [comment]
Probably not often, and judging from our unit tests we are even forgetting it exists in a lot of cases. It is slightly out of scope for this PR though (since it exists on RealmQuery), and we should probably benchmark it before removing it though?

File: realm/realm-library/src/main/java/io/realm/DynamicRealm.java [hunk 8]
```diff
@@ -167,28 +167,17 @@ public void executeTransaction(Transaction transaction) {
     }
 
     /**
-     * Gets all objects of a specific class name.
-     *
-     * @param className the Class to get objects of.
-     * @return a {@link RealmResults} list containing the objects. If no results where found, an empty list
-     * will be returned.
-     * @see io.realm.RealmResults
+     * DEPRECATED: Use {@code realm.where(className).findAll()} instead.
```
Comment (beeender) [inline realm/realm-library/src/main/java/io/realm/DynamicRealm.java]: `realm` - > 'dynamicRealm` ? Same for belows

File: realm/realm-library/src/androidTest/java/io/realm/DynamicRealmTests.java [hunk 9]
```diff
@@ -287,392 +278,7 @@ public void execute(DynamicRealm realm) {
         } finally {
             RealmLog.remove(testLogger);
         }
-        assertEquals(0, realm.allObjects("Owner").size());
-    }
-
-    @Test
-    public void allObjectsSorted() {
```
Comment (cmelchior) [inline realm/realm-library/src/androidTest/java/io/realm/DynamicRealmTests.java]: there is also `where()` test, so it should be safe to delete this IMO

**Comment by cmelchior** [comment]
All comments addressed @realm/java

File: realm/realm-library/src/androidTest/java/io/realm/DynamicRealmTests.java (hunk 9 above)
Comment (zaki50) [inline realm/realm-library/src/androidTest/java/io/realm/DynamicRealmTests.java]: which test? I could only find tests for `findAllSortedAsync()` not `findAllSorted()`.

**Comment by beeender** [comment]
:+1:

**Comment by stk1m1** [comment]
👍

**Pull request merged**